### PR TITLE
[FIX] account,*: remove manual use of check_move_validity

### PR DESCRIPTION
**Pull request by william-andre**
TLDR: the context key `check_move_validity` should never be used explicitly.

`check_move_validity` is a context key used to disable some important integrity check in accounting: `debit=credit`, which is one of the most fundamental rule.
It is possible to disable this in some cases, when creating and updating lines one by one, since the constraint cannot be satisfied between each line if the operations are not atomic.

A mechanism has been created for this, with `_check_balanced`. It is a context manager using `_disable_recursion`, which will taint all the contexts while inside of the context manager.
Putting any code changing lines one by one inside of that context manager will then allow having a transient invalid state, but still always check the validity at the end, when exiting the context.

That context key has been kept only in `point_of_sale` because a special wizard is used there to be able to validate a session with an unbalanced move, but a wizard is then in charge of balancing it. The check is then done at that time.

### CODE REVIEW

**Comment by william-andre** [comment]
For the reviewer: don't forget to use the "hide whitespace" option in Github's diff.
Some of the changes are done because pylint (and myself) wasn't happy with missing or useless spaces around operators

File: addons/account/models/account_account.py [hunk 1]
```diff
@@ -478,44 +478,52 @@ def _set_opening_debit_credit(self, amount, field):
...
+                    op_aml_credit = sum(account_op_lines.mapped('credit'))
+
+                    # There might be more than one line on this account if the opening entry was manually edited
+                    # If so, we need to merge all those lines into one before modifying its balance
+                    opening_move_line = account_op_lines[0]
+                    if len(account_op_lines) > 1:
+                        merge_write_cmd = [Command.update(opening_move_line.id, {
+                            'debit': op_aml_debit,
+                            'credit': op_aml_credit,
+                            'partner_id': None,
```
Comment (RicGR98) [inline addons/account/models/account_account.py:501]: Why `None` instead of `False` ? I see it was already there, but might as well clean it for better cohesion ?

**Review by RicGR98 — COMMENTED** [review]
Thanks for the interesting PR :)

There is still one useless `check_move_validity` in `test_unlink_timesheet`

File: addons/account/models/account_account.py (hunk 1 above)
Comment (william-andre) [reply addons/account/models/account_account.py]: > There is still one useless `check_move_validity` in `test_unlink_timesheet`

Didn't want to do diff in the test :shrug:

File: addons/point_of_sale/models/pos_order.py [hunk 2]
```diff
@@ -600,61 +600,62 @@ def _create_invoice(self, move_vals):
...
+                    existing_terms_line = new_move.line_ids.filtered(
+                        lambda line: line.account_id.account_type in ('asset_receivable', 'liability_payable'))
+                    if existing_terms_line.debit > 0:
+                        existing_terms_line_new_val = float_round(
+                            existing_terms_line.debit + rounding_line_difference,
+                            precision_rounding=new_move.currency_id.rounding)
+                    else:
+                        existing_terms_line_new_val = float_round(
+                            -existing_terms_line.credit + rounding_line_difference,
+                            precision_rounding=new_move.currency_id.rounding)
```
Comment (Levizar) [inline addons/point_of_sale/models/pos_order.py:654]: unrelated to the pr itself but it triggers me :laughing: 
```suggestion
                    balance = existing_terms_line.debit or -existing_terms_line.credit
                    existing_terms_line_new_val = float_round(
                        balance + rounding_line_difference,
                        precision_rounding=new_move.currency_id.rounding)
```

File: addons/point_of_sale/models/pos_order.py
Comment (Levizar) [inline addons/point_of_sale/models/pos_order.py:1]: Side comment: I think it would be nice to have something similar  for the `skip_invoice_sync` but I see this context key used a lot too.

**Review by Levizar — COMMENTED** [review]
There isn't really anything to say, well done.

> Didn't want to do diff in the test

why?

File: addons/point_of_sale/models/pos_order.py
Comment (william-andre) [reply addons/point_of_sale/models/pos_order.py:1]: Usually it is used to really disable it... Not only delay it.
It is less annoying than disabling a very important constraint

File: addons/account/demo/account_demo.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def _post_load_demo_data(self, company=False):
             + self.ref('demo_move_auto_reconcile_7')
             + self.ref('demo_move_auto_reconcile_8')
             + self.ref('demo_move_auto_reconcile_9')
-        ).with_context(check_move_validity=False)
```
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/demo/account_demo.py:50]: yes

**Review by Levizar — APPROVED** [review]
@robodoo r+

**Comment by Levizar** [comment]
@robodoo r+

**Review by Levizar — APPROVED** [review]
@robodoo r+
